src/ipred.rs

```diff
@@ -6,3 +12,167 @@ pub unsafe extern "C" fn get_upsample(
+        *const DynPixel,
+        libc::c_int,
+        libc::c_int,
+        *const int16_t,
+        libc::c_int,
+        libc::c_int,
+    ) -> (),
+>;
+pub type pal_pred_fn = Option<
+    unsafe extern "C" fn(
+        *mut DynPixel,
+        ptrdiff_t,
+        *const uint16_t,
+        *const uint8_t,
+        libc::c_int,
+        libc::c_int,
+    ) -> (),
+>;
+#[repr(C)]
+pub struct Dav1dIntraPredDSPContext {
+    pub intra_pred: [angular_ipred_fn; 14],
+    pub cfl_ac: [cfl_ac_fn; 3],
+    pub cfl_pred: [cfl_pred_fn; 6],
+    pub pal_pred: pal_pred_fn,
+}
+
+// TODO(legare): Generated fns are temporarily pub until init fns are deduplicated.
+#[cfg(feature = "asm")]
+macro_rules! decl_fn {
+    (angular_ipred, $name:ident) => {
+        pub(crate) fn $name(
+            dst: *mut DynPixel,
+            stride: ptrdiff_t,
+            topleft: *const DynPixel,
+            width: libc::c_int,
+            height: libc::c_int,
+            angle: libc::c_int,
+            max_width: libc::c_int,
+            max_height: libc::c_int,
+            bitdepth_max: libc::c_int,
+        );
+    };
+
+    (cfl_pred, $name:ident) => {
+        pub(crate) fn $name(
+            dst: *mut DynPixel,
+            stride: ptrdiff_t,
+            topleft: *const DynPixel,
+            width: libc::c_int,
+            height: libc::c_int,
+            ac: *const int16_t,
+            alpha: libc::c_int,
+            bitdepth_max: libc::c_int,
+        );
+    };
+
+    (cfl_ac, $name:ident) => {
+        pub(crate) fn $name(
+            ac: *mut int16_t,
+            y: *const DynPixel,
+            stride: ptrdiff_t,
+            w_pad: libc::c_int,
+            h_pad: libc::c_int,
+            cw: libc::c_int,
+            ch: libc::c_int,
+        );
+    };
+
+    (pal_pred, $name:ident) => {
+        pub(crate) fn $name(
+            dst: *mut DynPixel,
+            stride: ptrdiff_t,
+            pal: *const uint16_t,
+            idx: *const uint8_t,
+            w: libc::c_int,
+            h: libc::c_int,
+        );
+    };
+}
+
+#[cfg(feature = "asm")]
+macro_rules! decl_fns {
+    ($fn_kind:ident, $name:ident) => {
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        decl_fns!($fn_kind, $name, ssse3);
+
+        #[cfg(target_arch = "x86_64")]
+        decl_fns!($fn_kind, $name, avx2);
+
+        #[cfg(target_arch = "x86_64")]
+        decl_fns!($fn_kind, $name, avx512icl);
+
+        #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+        decl_fns!($fn_kind, $name, neon);
+    };
+
+    ($fn_kind:ident, $name:ident, $suffix:ident) => {
+        paste::paste! {
+            #[cfg(feature = "bitdepth_8")]
+            decl_fn!($fn_kind, [<dav1d_ $name _8bpc_ $suffix>]);
+            #[cfg(feature = "bitdepth_16")]
+            decl_fn!($fn_kind, [<dav1d_ $name _16bpc_ $suffix>]);
+        }
+    };
+}
+
+#[cfg(feature = "asm")]
+#[allow(dead_code)] // Macro declares more fns than actually exist.
```

Comment:
Could you do the same as https://github.com/memorysafety/rav1d/pull/416#discussion_r1316889886?